File: samples/py/src/rde_py/test/test_publisher.py
#!/usr/bin/env python3

import unittest
import pytest
import time

import rclpy
from rclpy.node import Node
from std_msgs.msg import String


class TestRdePublisher(unittest.TestCase):
    """
    Test suite for the RDE Publisher node.
    """

    @classmethod
    def setUpClass(cls):
        """Initialize ROS 2 for the test suite."""
        rclpy.init()

    @classmethod
    def tearDownClass(cls):
        """Shutdown ROS 2 after all tests."""
        rclpy.shutdown()

    def setUp(self):
        """Set up test fixtures."""
        self.node = rclpy.create_node('test_rde_publisher')
        self.received_messages = []

    def tearDown(self):
        """Clean up test fixtures."""
        self.node.destroy_node()

    def message_callback(self, msg):
        """Callback for receiving messages."""
        self.received_messages.append(msg.data)

    def test_publisher_exists(self):
        """Test that the publisher node exists and is publishing on /rde topic."""
        # Wait for publisher to appear
        start_time = time.time()
        publisher_found = False
        
        while time.time() - start_time < 5.0:
            topic_info = self.node.get_publishers_info_by_topic('/rde')
            if len(topic_info) > 0:
                publisher_found = True
                break
            time.sleep(0.1)
        
        if not publisher_found:
            pytest.skip("Publisher not found - rde_publisher node may not be running")
        
        self.assertTrue(publisher_found)

    def test_message_publishing(self):
        """Test that messages are being published on the /rde topic."""
        # Create a subscription
        subscription = self.node.create_subscription(
            String,
            '/rde',
            self.message_callback,
            10
        )
        
        # Spin for a few seconds to receive messages
        start_time = time.time()
        while time.time() - start_time < 3.0:
            rclpy.spin_once(self.node, timeout_sec=0.1)
        
        # Clean up
        self.node.destroy_subscription(subscription)
        
        # We should have received at least 2 messages (publisher runs at 1 Hz)
        if len(self.received_messages) == 0:
            pytest.skip("No messages received - rde_publisher node may not be running")
        
        self.assertGreaterEqual(len(self.received_messages), 2)
        
        # Verify message format
        for msg in self.received_messages:
            self.assertIn('Hello from RDE', msg)

    def test_message_increment(self):
        """Test that the message counter increments."""
        # Create a subscription
        subscription = self.node.create_subscription(
            String,
            '/rde',
            self.message_callback,
            10
        )
        
        # Spin for a few seconds
        start_time = time.time()
        while time.time() - start_time < 3.0:
            rclpy.spin_once(self.node, timeout_sec=0.1)
        
        # Clean up
        self.node.destroy_subscription(subscription)
        
        if len(self.received_messages) < 2:
            pytest.skip("Not enough messages received")
        
        # Messages should be different (counter increments)
        self.assertNotEqual(self.received_messages[0], self.received_messages[1])


if __name__ == '__main__':
    unittest.main()

File: samples/cpp/src/rde_cpp/test/test_publisher.cpp
#include <gtest/gtest.h>
#include <memory>
#include <chrono>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

using namespace std::chrono_literals;

class TestRdePublisher : public ::testing::Test
{
protected:
  static void SetUpTestCase()
  {
    rclcpp::init(0, nullptr);
  }

  static void TearDownTestCase()
  {
    rclcpp::shutdown();
  }

  void SetUp() override
  {
    node_ = std::make_shared<rclcpp::Node>("test_rde_publisher");
    executor_ = std::make_shared<rclcpp::executors::SingleThreadedExecutor>();
    executor_->add_node(node_);
    message_count_ = 0;
    received_messages_.clear();
  }

  void TearDown() override
  {
    executor_->cancel();
    executor_.reset();
    subscription_.reset();
    node_.reset();
  }

  void message_callback(const std_msgs::msg::String::SharedPtr msg)
  {
    message_count_++;
    received_messages_.push_back(msg->data);
  }

  std::shared_ptr<rclcpp::Node> node_;
  std::shared_ptr<rclcpp::executors::SingleThreadedExecutor> executor_;
  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription_;
  size_t message_count_;
  std::vector<std::string> received_messages_;
};

TEST_F(TestRdePublisher, TestPublisherExists)
{
  // Wait for the publisher to appear
  auto start_time = std::chrono::steady_clock::now();
  bool publisher_found = false;
  
  while (std::chrono::steady_clock::now() - start_time < 5s) {
    auto topic_info = node_->get_publishers_info_by_topic("/rde");
    if (!topic_info.empty()) {
      publisher_found = true;
      break;
    }
    std::this_thread::sleep_for(100ms);
  }
  
  if (!publisher_found) {
    GTEST_SKIP() << "Publisher not found - rde_publisher node may not be running";
  }
  
  EXPECT_TRUE(publisher_found);
}

TEST_F(TestRdePublisher, TestMessagePublishing)
{
  // Create a subscription to the /rde topic
  subscription_ = node_->create_subscription<std_msgs::msg::String>(
    "/rde",
    10,
    [this](const std_msgs::msg::String::SharedPtr msg) {
      this->message_callback(msg);
    }
  );
  
  // Spin for a few seconds to receive messages
  auto start_time = std::chrono::steady_clock::now();
  while (std::chrono::steady_clock::now() - start_time < 3s) {
    executor_->spin_some();
    std::this_thread::sleep_for(100ms);
  }
  
  // We should have received at least 2 messages (publisher runs at 1 Hz)
  if (message_count_ == 0) {
    GTEST_SKIP() << "No messages received - rde_publisher node may not be running";
  }
  
  EXPECT_GE(message_count_, 2);
  
  // Verify message format
  for (const auto& msg : received_messages_) {
    EXPECT_FALSE(msg.empty());
    EXPECT_NE(msg.find("hello rde"), std::string::npos);
  }
}

TEST_F(TestRdePublisher, TestMessageIncrement)
{
  // Create a subscription to the /rde topic
  subscription_ = node_->create_subscription<std_msgs::msg::String>(
    "/rde",
    10,
    [this](const std_msgs::msg::String::SharedPtr msg) {
      this->message_callback(msg);
    }
  );
  
  // Spin for a few seconds to receive messages
  auto start_time = std::chrono::steady_clock::now();
  while (std::chrono::steady_clock::now() - start_time < 3s) {
    executor_->spin_some();
    std::this_thread::sleep_for(100ms);
  }
  
  if (received_messages_.size() < 2) {
    GTEST_SKIP() << "Not enough messages received";
  }
  
  // The messages should contain incrementing numbers
  // This is a basic check that the counter is working
  EXPECT_NE(received_messages_[0], received_messages_[1]);
}

int main(int argc, char ** argv)
{
  ::testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}
